python/lsst/ip/diffim/imageMapReduce.py

```diff
@@ -0,0 +1,575 @@
+        """Perform `mapperSubtask.run()` on each patch
+
+        Perform `mapperSubtask.run()` on each patch across a
+        grid on `exposure` generated by `generateGrid`.
+
+        Parameters
+        ----------
+        exposure : afwImage.ExposureF
+            the original exposure which is used as the template
+        doClone : boolean
+            if True, clone the subimages before passing to subtask;
+            in that case, the sub-exps do not have to be considered as read-only
+        kwargs :
+            additional keyword arguments to be passed to
+            `mapperSubtask.run`
+
+        Returns
+        -------
+        a list of `afwExposure`s or other values returned by `mapperSubtask.run`.
+        """
+        boxes0, boxes1 = self.boxes0, self.boxes1
+        if boxes0 is None:
+            boxes0, boxes1 = self._generateGrid(exposure)
+        if len(boxes0) != len(boxes1):
+            raise Exception('Uh oh!')   # TBD: define a specific exception to raise
+
+        self.log.info("Processing %d sub-exposures" % len(boxes0))
+        patches = []
+        for i in range(len(boxes0)):
+            subExp = afwImage.ExposureF(exposure, boxes0[i])
+            expandedSubExp = afwImage.ExposureF(exposure, boxes1[i])
+            if doClone:
+                subExp = subExp.clone()
+                expandedSubExp = expandedSubExp.clone()
+            result = self.mapperSubtask.run(subExp, expandedSubExp, exposure.getBBox(), **kwargs)
+            patches.append(result)
+
+        return patches
+
+    def _reduceImage(self, patches, exposure, **kwargs):
+        """Reduce/merge a set of sub-exposure patches into a final exposure
+
+        Return an exposure of the same dimensions as `exposure`.
+        `patches` is expected to have been produced by `runMapper`.
+
+        Parameters
+        ----------
+        patches : list
+            list of subExposures of `exposure` to merge
+        exposure : afwImage.Exposure
+            the original exposure
+        **kwargs :
+            additional keyword arguments
+
+        Returns
+        -------
+        Output of `reducerSubtask.run` which may be an `afw.Exposure` or
+        another iterable.
+        """
+        result = self.reducerSubtask.run(patches, exposure, **kwargs)
+        return result
+
+    def _generateGrid(self, exposure, forceEvenSized=False):
+        """Generate two lists of bounding boxes that evenly grid `exposure`
+
+        Unless the config was provided with `centroidCoordsX` and
+        `centroidCoordsY`, grid (subimage) centers are spaced
+        evenly by gridStepX/Y. Then the grid is adjusted as
+        little as possible to evenly cover the input exposure (if
+        adjustGridOption is not 'none'). Then the second set of
+        bounding boxes is expanded by borderSizeX/Y. The expanded
+        bounding boxes are adjusted to ensure that they intersect
+        the exposure's bounding box. The resulting lists of bounding
```

Comment:
Do you need to both return the lists and save them as instance variables? I think the latter is more than sufficient, especially since this is a "private" method.